### PR TITLE
backend/feat/registry and gateway fallback logic

### PR DESCRIPTION
**Pull request by shreyaxh**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates

### Additional Changes

- [ ] This PR modifies the database schema (database migration added)
- [ ] This PR modifies dhall configs/environment variables
- [ ] This PR contains API breaking changes

## Checklist

- [ ] I formatted the code and addressed linter errors `./dev/format-all-files.sh`
- [ ] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added integration tests for my changes where possible
- [ ] I tested the changes using integration tests
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable
- [ ] No leak detected in leakcanary

### CODE REVIEW

File: Backend/app/dashboard/CommonAPIs/src/Dashboard/Common/Merchant.hs [hunk 1]
```diff
@@ -778,3 +779,71 @@ data VehicleClassVariantMap = VehicleClassVariantMap
...
+  deriving stock (Show, Eq, Ord, Read, Generic)
+  deriving anyclass (ToJSON, FromJSON, ToSchema, Kernel.Prelude.ToParamSchema)
+
+instance FromHttpApiData ConfigNames where
+  parseQueryParam = readEither
+
+instance ToHttpApiData ConfigNames where
+  toUrlPiece = show
+
+data SMSProviderEnums
```
Comment (NikithShetty) [inline Backend/app/dashboard/CommonAPIs/src/Dashboard/Common/Merchant.hs:806]: can we make use of existing enums for this?

File: Backend/app/rider-platform/rider-app/Main/spec/Storage/Merchant.yaml [hunk 2]
```diff
@@ -201,6 +207,7 @@ Merchant:
     fakeOtpEmails: "'{}'"
     city: "'Kochi'"
     onlinePayment: "false"
+    gatewayAndRegistryPriorityList: "'{\"ONDC\", \"NY\"}'"
```
Comment (NikithShetty) [inline Backend/app/rider-platform/rider-app/Main/spec/Storage/Merchant.yaml:210]: Let default be just NY

File: Backend/app/rider-platform/rider-app/Main/src/Environment.hs [hunk 3]
```diff
@@ -294,11 +303,26 @@ instance Registry Flow where
...
+        registryUrl <-
+          case merchant.gatewayAndRegistryPriorityList of
+            (NY : _) -> asks (.nyRegistryUrl)
+            _ -> asks (.ondcRegistryUrl)
+        pure registryUrl
+      handleRegistryError :: ExternalAPICallError -> BaseUrl -> SimpleLookupRequest -> Flow (Maybe Subscriber)
+      handleRegistryError _ registryUrl sub = do
+        merchant <- CM.findById (Id sub.merchant_id) >>= fromMaybeM (MerchantDoesNotExist sub.merchant_id)
+        let networkPriorityList = reorderList merchant.gatewayAndRegistryPriorityList
+        CM.updateGatewayAndRegistryPriorityList merchant networkPriorityList
```
Comment (NikithShetty) [inline Backend/app/rider-platform/rider-app/Main/src/Environment.hs:321]: Let's not have this logic here. We can make use of failover API to do the switching which can be called based on prom metrics.

File: Backend/dhall-configs/dev/common.dhall [hunk 4]
```diff
@@ -48,6 +48,14 @@ let SchedulerType = < RedisBased | DbBased >
 let urlShortnerConfig =
       { url = "http://localhost:9023/", apiKey = sec.urlShortnerApiKey }
 
+let ondcRegistryUrl = "ondcRegistryUrl"
+
+let ondcGatewayUrl = "ondcGatewayUrl"
+
+let nyRegistryUrl = "nyRegistryUrl"
```
Comment (NikithShetty) [inline Backend/dhall-configs/dev/common.dhall:55]: you need to update with actual localhost url here, else dev testing will fail

**Review by NikithShetty — CHANGES_REQUESTED** [review]

File: Backend/app/rider-platform/rider-app/Main/src/Environment.hs (hunk 3 above)
Comment (shreyaxh) [reply Backend/app/rider-platform/rider-app/Main/src/Environment.hs:321]: Here should we call performLookup withRetryConfig and check with the other registry?
And then call the update function based on metrics

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Cac/MerchantServiceUsageConfig.hs [hunk 5]
```diff
@@ -67,5 +67,9 @@ clearCache = CMMSUC.clearCache
 updateMerchantServiceUsageConfig :: (MonadFlow m, CacheFlow m r, EsqDBFlow m r) => MerchantServiceUsageConfig -> m ()
 updateMerchantServiceUsageConfig = CMMSUC.updateMerchantServiceUsageConfig
 
+-- updateSmsProvidersPriorityList,
```
Comment (NikithShetty) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Cac/MerchantServiceUsageConfig.hs:70]: Remove commented out code

File: Backend/app/rider-platform/rider-app/Main/src/SharedLogic/CallBPP.hs [hunk 6]
```diff
@@ -66,6 +66,17 @@ searchV2 gatewayUrl req merchantId = do
   internalEndPointHashMap <- asks (.internalEndPointHashMap)
   bapId <- req.searchReqContext.contextBapId & fromMaybeM (InvalidRequest "BapId is missing")
   callBecknAPIWithSignature' merchantId bapId "search" API.searchAPIV2 gatewayUrl internalEndPointHashMap req
+    `catch` \e -> handleGatewayError e
+  where
+    handleGatewayError :: (MonadFlow m, CacheFlow m r, EsqDBFlow m r) => ExternalAPICallError -> m AckResponse
+    handleGatewayError _ = do
+      merchant <- CQM.findById merchantId >>= fromMaybeM (MerchantDoesNotExist merchantId.getId)
+      let networkPriorityList = reorderList merchant.gatewayAndRegistryPriorityList
+      CQM.updateGatewayAndRegistryPriorityList merchant networkPriorityList
```
Comment (NikithShetty) [inline Backend/app/rider-platform/rider-app/Main/src/SharedLogic/CallBPP.hs:75]: Don't update the list in DB, just use the next gateway url at runtime

**Review by NikithShetty — CHANGES_REQUESTED** [review]

**Review by NikithShetty — APPROVED** [review]

**Review by prakharritik — CHANGES_REQUESTED** [review]
Squash the commits or change the commit message of any.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Dashboard/Management/Merchant.hs [hunk 7]
```diff
@@ -2259,3 +2260,54 @@ postMerchantConfigClearCacheSubscription merchantShortId opCity req = do
...
+          whatsappProviderPriorityList = reorderList merchantServiceUsageConfig.whatsappProvidersPriorityList
+      let updatedConfig = merchantServiceUsageConfig {DMSUC.smsProvidersPriorityList = messageProviderPriorityList, DMSUC.whatsappProvidersPriorityList = whatsappProviderPriorityList}
+      CQMSUC.updateMerchantServiceUsageConfig updatedConfig
+  pure ()
+
+reorderList :: [a] -> [a]
+reorderList [] = []
+reorderList (x : xs) = xs ++ [x]
+
+notEmpty :: [a] -> Bool
```
Comment (prakharritik) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Dashboard/Management/Merchant.hs:2307]: we can use not null

**Review by prakharritik — APPROVED** [review]

**Pull request merged**